File: go/interface-segregation/ispCorrectDemonstration.go
// Interface Segregation Principle (ISP) - CORRECT Example (Go)
//
// This example demonstrates proper interface segregation where clients
// only depend on the methods they actually need.
//
// Solution: Break down the monolithic Human interface into smaller,
// focused interfaces that can be implemented independently.

package main

import (
	"errors"
	"fmt"
)

// CORRECT: Small, focused interfaces following ISP
type Workable interface {
	Work()
}

type Biological interface {
	Eat() error
	Sleep() error
	Breathe() error
}

type Cognitive interface {
	Think()
	Learn() error
}

type Social interface {
	Socialize()
	Communicate()
}

type Physical interface {
	Exercise() error
	Move()
}

type Reproductive interface {
	Reproduce() error
}

// Robot-specific interfaces
type Mechanical interface {
	PowerDown()
	PowerUp()
	RunDiagnostics()
}

type Rechargeable interface {
	Recharge()
	GetBatteryLevel() int
}

type Programmable interface {
	LoadProgram(program string)
	ExecuteProgram() error
}

// Advanced robot capabilities
type NetworkEnabled interface {
	ConnectToNetwork() error
	SendData(data string) error
	ReceiveData() (string, error)
}

// Employee implementation - implements relevant interfaces
type Employee struct {
	Name        string
	Energy      int
	IsAwake     bool
	CurrentTask string
}

func NewEmployee(name string) *Employee {
	return &Employee{
		Name:    name,
		Energy:  100,
		IsAwake: true,
	}
}

// Workable interface
func (e *Employee) Work() {
	if !e.IsAwake {
		fmt.Printf("%s cannot work while sleeping!\n", e.Name)
		return
	}
	task := e.CurrentTask
	if task == "" {
		task = "general tasks"
	}
	fmt.Printf("%s is working on: %s\n", e.Name, task)
	e.Energy -= 20
}

// Biological interface
func (e *Employee) Eat() error {
	fmt.Printf("%s is eating nutritious food\n", e.Name)
	e.Energy += 30
	return nil
}

func (e *Employee) Sleep() error {
	fmt.Printf("%s is sleeping peacefully\n", e.Name)
	e.IsAwake = false
	e.Energy = 100
	return nil
}

func (e *Employee) Breathe() error {
	fmt.Printf("%s is breathing fresh air\n", e.Name)
	return nil
}

// Cognitive interface
func (e *Employee) Think() {
	fmt.Printf("%s is thinking creatively and analytically\n", e.Name)
	e.Energy -= 5
}

func (e *Employee) Learn() error {
	fmt.Printf("%s is learning new skills\n", e.Name)
	e.Energy -= 10
	return nil
}

// Social interface
func (e *Employee) Socialize() {
	fmt.Printf("%s is socializing with friends\n", e.Name)
	e.Energy -= 10
}

func (e *Employee) Communicate() {
	fmt.Printf("%s is communicating through speech\n", e.Name)
}

// Physical interface
func (e *Employee) Exercise() error {
	fmt.Printf("%s is doing physical exercise\n", e.Name)
	e.Energy -= 15
	return nil
}

func (e *Employee) Move() {
	fmt.Printf("%s is walking around\n", e.Name)
	e.Energy -= 5
}

// Reproductive interface
func (e *Employee) Reproduce() error {
	fmt.Printf("%s can participate in human reproduction\n", e.Name)
	return nil
}

func (e *Employee) SetCurrentTask(task string) {
	e.CurrentTask = task
}

// FactoryBot - only implements interfaces it needs
type FactoryBot struct {
	Model          string
	BatteryLevel   int
	IsOperational  bool
	CurrentProgram string
}

func NewFactoryBot(model string) *FactoryBot {
	return &FactoryBot{
		Model:         model,
		BatteryLevel:  100,
		IsOperational: true,
	}
}

// Workable interface
func (r *FactoryBot) Work() {
	if !r.IsOperational {
		fmt.Printf("%s is not operational!\n", r.Model)
		return
	}
	if r.CurrentProgram == "" {
		fmt.Printf("%s has no program loaded!\n", r.Model)
		return
	}
	fmt.Printf("%s robot is executing work program: %s\n", r.Model, r.CurrentProgram)
	r.BatteryLevel -= 10
}

// Mechanical interface
func (r *FactoryBot) PowerDown() {
	fmt.Printf("%s is powering down all systems\n", r.Model)
	r.IsOperational = false
}

func (r *FactoryBot) PowerUp() {
	fmt.Printf("%s is powering up systems\n", r.Model)
	r.IsOperational = true
}

func (r *FactoryBot) RunDiagnostics() {
	fmt.Printf("%s is running comprehensive diagnostics\n", r.Model)
	status := "Offline"
	if r.IsOperational {
		status = "Operational"
	}
	fmt.Printf("Battery: %d%%, Status: %s\n", r.BatteryLevel, status)
}

// Rechargeable interface
func (r *FactoryBot) Recharge() {
	fmt.Printf("%s is recharging battery to full capacity\n", r.Model)
	r.BatteryLevel = 100
}

func (r *FactoryBot) GetBatteryLevel() int {
	return r.BatteryLevel
}

// Programmable interface
func (r *FactoryBot) LoadProgram(program string) {
	fmt.Printf("%s is loading program: %s\n", r.Model, program)
	r.CurrentProgram = program
}

func (r *FactoryBot) ExecuteProgram() error {
	if r.CurrentProgram == "" {
		return errors.New("no program loaded")
	}
	fmt.Printf("%s is executing program: %s\n", r.Model, r.CurrentProgram)
	r.BatteryLevel -= 5
	return nil
}

// AndroidBot - implements more interfaces including some human-like ones
type AndroidBot struct {
	Name             string
	BatteryLevel     int
	IsOperational    bool
	CurrentProgram   string
	NetworkConnected bool
}

func NewAndroidBot(name string) *AndroidBot {
	return &AndroidBot{
		Name:             name,
		BatteryLevel:     100,
		IsOperational:    true,
		NetworkConnected: false,
	}
}

// Workable interface
func (h *AndroidBot) Work() {
	if !h.IsOperational {
		fmt.Printf("%s is not operational!\n", h.Name)
		return
	}
	fmt.Printf("%s humanoid robot is performing complex work tasks\n", h.Name)
	h.BatteryLevel -= 8
}

// Cognitive interface - robots can have AI cognition
func (h *AndroidBot) Think() {
	fmt.Printf("%s is processing information using AI algorithms\n", h.Name)
	h.BatteryLevel -= 3
}

func (h *AndroidBot) Learn() error {
	fmt.Printf("%s is updating neural networks with new data\n", h.Name)
	h.BatteryLevel -= 5
	return nil
}

// Social interface - humanoid robots can be social
func (h *AndroidBot) Socialize() {
	fmt.Printf("%s is engaging in social interaction protocols\n", h.Name)
	h.BatteryLevel -= 4
}

func (h *AndroidBot) Communicate() {
	fmt.Printf("%s is communicating through speech synthesis\n", h.Name)
}

// Physical interface - humanoid robots can move
func (h *AndroidBot) Move() {
	fmt.Printf("%s is walking with bipedal locomotion\n", h.Name)
	h.BatteryLevel -= 6
}

func (h *AndroidBot) Exercise() error {
	fmt.Printf("%s is performing calibration movements\n", h.Name)
	h.BatteryLevel -= 7
	return nil
}

// Mechanical interface
func (h *AndroidBot) PowerDown() {
	fmt.Printf("%s humanoid is entering sleep mode\n", h.Name)
	h.IsOperational = false
}

func (h *AndroidBot) PowerUp() {
	fmt.Printf("%s humanoid is activating all systems\n", h.Name)
	h.IsOperational = true
}

func (h *AndroidBot) RunDiagnostics() {
	fmt.Printf("%s is running humanoid system diagnostics\n", h.Name)
	network := "Disconnected"
	if h.NetworkConnected {
		network = "Connected"
	}
	fmt.Printf("Battery: %d%%, Network: %s\n", h.BatteryLevel, network)
}

// Rechargeable interface
func (h *AndroidBot) Recharge() {
	fmt.Printf("%s is recharging via induction pad\n", h.Name)
	h.BatteryLevel = 100
}

func (h *AndroidBot) GetBatteryLevel() int {
	return h.BatteryLevel
}

// Programmable interface
func (h *AndroidBot) LoadProgram(program string) {
	fmt.Printf("%s is loading behavioral program: %s\n", h.Name, program)
	h.CurrentProgram = program
}

func (h *AndroidBot) ExecuteProgram() error {
	if h.CurrentProgram == "" {
		return errors.New("no behavioral program loaded")
	}
	fmt.Printf("%s is executing behavioral program: %s\n", h.Name, h.CurrentProgram)
	h.BatteryLevel -= 3
	return nil
}

// NetworkEnabled interface
func (h *AndroidBot) ConnectToNetwork() error {
	fmt.Printf("%s is connecting to robot network\n", h.Name)
	h.NetworkConnected = true
	return nil
}

func (h *AndroidBot) SendData(data string) error {
	if !h.NetworkConnected {
		return errors.New("not connected to network")
	}
	fmt.Printf("%s is sending data: %s\n", h.Name, data)
	return nil
}

func (h *AndroidBot) ReceiveData() (string, error) {
	if !h.NetworkConnected {
		return "", errors.New("not connected to network")
	}
	fmt.Printf("%s is receiving network data\n", h.Name)
	return "network_data_received", nil
}

// Client classes that only depend on interfaces they need
type TaskManager struct{}

// Only depends on Workable interface - works with any worker type
func (tm *TaskManager) AssignTask(worker Workable, task string) {
	fmt.Printf("Assigning task: %s\n", task)

	// Type-specific task assignment
	switch w := worker.(type) {
	case *Employee:
		w.SetCurrentTask(task)
	case Programmable:
		w.LoadProgram(task)
	}

	worker.Work()
}

// Only needs workers, doesn't care about other capabilities
func (tm *TaskManager) ManageWorkforce(workers ...Workable) {
	fmt.Printf("Managing workforce of %d workers\n", len(workers))
	for _, worker := range workers {
		worker.Work()
	}
}

type TechnicianManager struct{}

// Only depends on Mechanical interface
func (tm *TechnicianManager) PerformMaintenance(device Mechanical) {
	fmt.Println("Starting maintenance routine...")
	device.RunDiagnostics()
	device.PowerDown()
	fmt.Println("Maintenance completed")
	device.PowerUp()
}

// Only depends on Rechargeable interface
func (tm *TechnicianManager) ManageCharge(device Rechargeable) {
	if device.GetBatteryLevel() < 20 {
		fmt.Println("Battery low, initiating recharge")
		device.Recharge()
	} else {
		fmt.Printf("Battery sufficient: %d%%\n", device.GetBatteryLevel())
	}
}

type TrainingManager struct{}

// Only depends on Cognitive interface
func (tm *TrainingManager) ConductTraining(learner Cognitive) {
	fmt.Println("Starting cognitive training session...")
	learner.Think()
	if err := learner.Learn(); err != nil {
		fmt.Printf("Learning error: %s\n", err.Error())
	}
}

type EventCoordinator struct{}

// Only depends on Social interface
func (ec *EventCoordinator) OrganizeSocialEvent(participants ...Social) {
	fmt.Println("Organizing social interaction...")
	for _, participant := range participants {
		participant.Socialize()
		participant.Communicate()
	}
}

type HealthCareProvider struct{}

// Only depends on Biological interface
func (hcp *HealthCareProvider) ProvideCare(organism Biological) error {
	fmt.Println("Providing biological care...")

	if err := organism.Eat(); err != nil {
		return fmt.Errorf("feeding failed: %w", err)
	}

	if err := organism.Sleep(); err != nil {
		return fmt.Errorf("rest failed: %w", err)
	}

	if err := organism.Breathe(); err != nil {
		return fmt.Errorf("breathing failed: %w", err)
	}

	return nil
}

// Demonstrates proper Interface Segregation
func mainCorrect() {
	fmt.Println("=== Interface Segregation Principle - CORRECT (Go) ===")
	fmt.Println()

	// Create different types of workers
	alice := NewEmployee("Alice")
	factoryBot := NewFactoryBot("FactoryBot-3000")
	android := NewAndroidBot("Android-Sara")

	tm := &TaskManager{}
	techMgr := &TechnicianManager{}
	trainMgr := &TrainingManager{}
	eventCoord := &EventCoordinator{}
	healthProv := &HealthCareProvider{}

	fmt.Println("1. Work Management (only needs Workable interface):")
	tm.AssignTask(alice, "Design new product")
	tm.AssignTask(factoryBot, "Assembly line task")
	tm.AssignTask(android, "Customer service")

	fmt.Println()
	fmt.Println("2. Managing workforce (polymorphic with Workable):")
	tm.ManageWorkforce(alice, factoryBot, android)

	fmt.Println()
	fmt.Println("3. Maintenance (only for Mechanical devices):")
	// Note: Cannot pass alice here - she doesn't implement Mechanical!
	techMgr.PerformMaintenance(factoryBot)
	techMgr.PerformMaintenance(android)

	fmt.Println()
	fmt.Println("4. Battery Management (only for Rechargeable devices):")
	// Note: Cannot pass alice here - she doesn't implement Rechargeable!
	techMgr.ManageCharge(factoryBot)
	techMgr.ManageCharge(android)

	fmt.Println()
	fmt.Println("5. Cognitive Training (for thinking entities):")
	trainMgr.ConductTraining(alice)
	trainMgr.ConductTraining(android) // Humanoid can think too!
	// Note: Cannot pass factoryBot - it doesn't implement Cognitive!

	fmt.Println()
	fmt.Println("6. Social Coordination (for social entities):")
	eventCoord.OrganizeSocialEvent(alice, android)
	// Note: Cannot pass factoryBot - it doesn't implement Social!

	fmt.Println()
	fmt.Println("7. Biological Care (only for living organisms):")
	if err := healthProv.ProvideCare(alice); err != nil {
		fmt.Printf("Care failed: %s\n", err.Error())
	}
	// Note: Cannot pass robots - they don't implement Biological!

	fmt.Println()
	fmt.Println("8. Network Operations (only for NetworkEnabled):")
	// Use type assertion to check if android implements NetworkEnabled
	if netDevice, ok := interface{}(android).(NetworkEnabled); ok {
		if err := netDevice.ConnectToNetwork(); err == nil {
			netDevice.SendData("Status update")
			_, _ = netDevice.ReceiveData()
		}
	}

	fmt.Println()
	fmt.Println("=== Benefits of Interface Segregation ===")
	fmt.Println("✓ Clients only depend on methods they actually use")
	fmt.Println("✓ No forced implementation of irrelevant methods")
	fmt.Println("✓ High cohesion within each interface")
	fmt.Println("✓ Easy to extend with new capabilities")
	fmt.Println("✓ Better testability and maintainability")
	fmt.Println("✓ Follows Single Responsibility at interface level")
	fmt.Println("✓ Go's implicit interface satisfaction enables clean design")

	fmt.Println()
	fmt.Println("=== Interface Composition Examples ===")
	fmt.Println("Employee implements: Workable + Biological + Cognitive + Social + Physical + Reproductive")
	fmt.Println("FactoryBot implements: Workable + Mechanical + Rechargeable + Programmable")
	fmt.Println("AndroidBot implements: Workable + Cognitive + Social + Physical + Mechanical + Rechargeable + Programmable + NetworkEnabled")
}

func mainIncorrect() {
	// Run the correct demonstration
	mainCorrect()
}

File: javascript/interface-segregation/ispCorrectDemonstration.js
/**
 * Interface Segregation Principle (ISP) - CORRECT Example (JavaScript)
 * 
 * This example demonstrates proper interface segregation where clients
 * only depend on the methods they actually need.
 * 
 * Solution: Break down the monolithic Human "interface" into smaller,
 * focused contracts that can be implemented independently.
 */

/**
 * CORRECT: Small, focused "interfaces" (expected contracts) following ISP
 * In JavaScript, we simulate interfaces through duck typing and expected method contracts
 */

// Workable contract - anything that can perform work
const WorkableContract = {
    work: 'function'
};

// Biological contract - living organisms
const BiologicalContract = {
    eat: 'function',
    sleep: 'function',
    breathe: 'function'
};

// Cognitive contract - thinking entities
const CognitiveContract = {
    think: 'function',
    learn: 'function'
};

// Social contract - entities that can interact socially
const SocialContract = {
    socialize: 'function',
    communicate: 'function'
};

// Physical contract - entities that can move and exercise
const PhysicalContract = {
    exercise: 'function',
    move: 'function'
};

// Mechanical contract - machines with power and diagnostic capabilities
const MechanicalContract = {
    powerDown: 'function',
    powerUp: 'function',
    runDiagnostics: 'function'
};

// Rechargeable contract - devices with battery management
const RechargeableContract = {
    recharge: 'function',
    getBatteryLevel: 'function'
};

// Programmable contract - devices that can execute programs
const ProgrammableContract = {
    loadProgram: 'function',
    executeProgram: 'function'
};

// NetworkEnabled contract - devices with network capabilities
const NetworkEnabledContract = {
    connectToNetwork: 'function',
    sendData: 'function',
    receiveData: 'function'
};

/**
 * Helper function to check if object follows a contract
 */
function followsContract(obj, contract) {
    return Object.keys(contract).every(method => typeof obj[method] === contract[method]);
}

/**
 * Helper function to get contracts followed by an object
 */
function getFollowedContracts(obj) {
    const contracts = {
        Workable: WorkableContract,
        Biological: BiologicalContract,
        Cognitive: CognitiveContract,
        Social: SocialContract,
        Physical: PhysicalContract,
        Mechanical: MechanicalContract,
        Rechargeable: RechargeableContract,
        Programmable: ProgrammableContract,
        NetworkEnabled: NetworkEnabledContract
    };
    
    return Object.keys(contracts).filter(name => followsContract(obj, contracts[name]));
}

/**
 * Human implementation - implements relevant contracts
 */
class Person {
    constructor(name) {
        this.name = name;
        this.energy = 100;
        this.isAwake = true;
        this.currentTask = null;
    }
    
    // Workable contract
    work() {
        if (!this.isAwake) {
            console.log(`${this.name} cannot work while sleeping!`);
            return;
        }
        console.log(`${this.name} is working on: ${this.currentTask || 'general tasks'}`);
        this.energy -= 20;
    }
    
    // Biological contract
    eat() {
        console.log(`${this.name} is eating nutritious food`);
        this.energy += 30;
    }
    
    sleep() {
        console.log(`${this.name} is sleeping peacefully`);
        this.isAwake = false;
        this.energy = 100;
    }
    
    breathe() {
        console.log(`${this.name} is breathing fresh air`);
    }
    
    // Cognitive contract
    think() {
        console.log(`${this.name} is thinking creatively and analytically`);
        this.energy -= 5;
    }
    
    learn() {
        console.log(`${this.name} is learning new skills`);
        this.energy -= 10;
    }
    
    // Social contract
    socialize() {
        console.log(`${this.name} is socializing with friends`);
        this.energy -= 10;
    }
    
    communicate() {
        console.log(`${this.name} is communicating through speech`);
    }
    
    // Physical contract
    exercise() {
        console.log(`${this.name} is doing physical exercise`);
        this.energy -= 15;
    }
    
    move() {
        console.log(`${this.name} is walking around`);
        this.energy -= 5;
    }
    
    setCurrentTask(task) {
        this.currentTask = task;
    }
}

/**
 * Basic Robot - only implements contracts it needs
 */
class IndustrialRobot {
    constructor(model) {
        this.model = model;
        this.batteryLevel = 100;
        this.isOperational = true;
        this.currentProgram = null;
    }
    
    // Workable contract
    work() {
        if (!this.isOperational) {
            console.log(`${this.model} is not operational!`);
            return;
        }
        if (!this.currentProgram) {
            console.log(`${this.model} has no program loaded!`);
            return;
        }
        console.log(`${this.model} robot is executing work program: ${this.currentProgram}`);
        this.batteryLevel -= 10;
    }
    
    // Mechanical contract
    powerDown() {
        console.log(`${this.model} is powering down all systems`);
        this.isOperational = false;
    }
    
    powerUp() {
        console.log(`${this.model} is powering up systems`);
        this.isOperational = true;
    }
    
    runDiagnostics() {
        console.log(`${this.model} is running comprehensive diagnostics`);
        console.log(`Battery: ${this.batteryLevel}%, Status: ${this.isOperational ? 'Operational' : 'Offline'}`);
    }
    
    // Rechargeable contract
    recharge() {
        console.log(`${this.model} is recharging battery to full capacity`);
        this.batteryLevel = 100;
    }
    
    getBatteryLevel() {
        return this.batteryLevel;
    }
    
    // Programmable contract
    loadProgram(program) {
        console.log(`${this.model} is loading program: ${program}`);
        this.currentProgram = program;
    }
    
    executeProgram() {
        if (!this.currentProgram) {
            console.log(`${this.model} has no program to execute!`);
            return;
        }
        console.log(`${this.model} is executing program: ${this.currentProgram}`);
        this.batteryLevel -= 5;
    }
}

/**
 * Advanced Humanoid Robot - implements more contracts including some human-like ones
 */
class HumanoidRobot {
    constructor(name) {
        this.name = name;
        this.batteryLevel = 100;
        this.isOperational = true;
        this.currentProgram = null;
        this.networkConnected = false;
    }
    
    // Workable contract
    work() {
        if (!this.isOperational) {
            console.log(`${this.name} is not operational!`);
            return;
        }
        console.log(`${this.name} humanoid robot is performing complex work tasks`);
        this.batteryLevel -= 8;
    }
    
    // Cognitive contract - robots can have AI cognition
    think() {
        console.log(`${this.name} is processing information using AI algorithms`);
        this.batteryLevel -= 3;
    }
    
    learn() {
        console.log(`${this.name} is updating neural networks with new data`);
        this.batteryLevel -= 5;
    }
    
    // Social contract - humanoid robots can be social
    socialize() {
        console.log(`${this.name} is engaging in social interaction protocols`);
        this.batteryLevel -= 4;
    }
    
    communicate() {
        console.log(`${this.name} is communicating through speech synthesis`);
    }
    
    // Physical contract - humanoid robots can move
    move() {
        console.log(`${this.name} is walking with bipedal locomotion`);
        this.batteryLevel -= 6;
    }
    
    exercise() {
        console.log(`${this.name} is performing calibration movements`);
        this.batteryLevel -= 7;
    }
    
    // Mechanical contract
    powerDown() {
        console.log(`${this.name} humanoid is entering sleep mode`);
        this.isOperational = false;
    }
    
    powerUp() {
        console.log(`${this.name} humanoid is activating all systems`);
        this.isOperational = true;
    }
    
    runDiagnostics() {
        console.log(`${this.name} is running humanoid system diagnostics`);
        console.log(`Battery: ${this.batteryLevel}%, Network: ${this.networkConnected ? 'Connected' : 'Disconnected'}`);
    }
    
    // Rechargeable contract
    recharge() {
        console.log(`${this.name} is recharging via induction pad`);
        this.batteryLevel = 100;
    }
    
    getBatteryLevel() {
        return this.batteryLevel;
    }
    
    // Programmable contract
    loadProgram(program) {
        console.log(`${this.name} is loading behavioral program: ${program}`);
        this.currentProgram = program;
    }
    
    executeProgram() {
        if (!this.currentProgram) {
            console.log(`${this.name} has no behavioral program loaded!`);
            return;
        }
        console.log(`${this.name} is executing behavioral program: ${this.currentProgram}`);
        this.batteryLevel -= 3;
    }
    
    // NetworkEnabled contract
    connectToNetwork() {
        console.log(`${this.name} is connecting to robot network`);
        this.networkConnected = true;
    }
    
    sendData(data) {
        if (!this.networkConnected) {
            console.log(`${this.name} is not connected to network!`);
            return;
        }
        console.log(`${this.name} is sending data: ${data}`);
    }
    
    receiveData() {
        if (!this.networkConnected) {
            console.log(`${this.name} is not connected to network!`);
            return;
        }
        console.log(`${this.name} is receiving network data`);
    }
}

/**
 * Client classes that only depend on contracts they need
 */
class WorkManager {
    // Only depends on Workable contract - works with any worker type
    static assignTask(worker, task) {
        if (!followsContract(worker, WorkableContract)) {
            throw new Error("Worker must follow Workable contract");
        }
        
        console.log(`Assigning task: ${task}`);
        if (worker.setCurrentTask) {
            worker.setCurrentTask(task);
        } else if (followsContract(worker, ProgrammableContract)) {
            worker.loadProgram(task);
        }
        worker.work();
    }
    
    // Only needs workers, doesn't care about other capabilities
    static manageWorkforce(...workers) {
        console.log(`Managing workforce of ${workers.length} workers`);
        workers.forEach(worker => {
            if (followsContract(worker, WorkableContract)) {
                worker.work();
            } else {
                console.log("Warning: Worker doesn't follow Workable contract");
            }
        });
    }
}

class MaintenanceManager {
    // Only depends on Mechanical contract
    static performMaintenance(device) {
        if (!followsContract(device, MechanicalContract)) {
            throw new Error("Device must follow Mechanical contract");
        }
        
        console.log("Starting maintenance routine...");
        device.runDiagnostics();
        device.powerDown();
        console.log("Maintenance completed");
        device.powerUp();
    }
    
    // Only depends on Rechargeable contract
    static manageCharge(device) {
        if (!followsContract(device, RechargeableContract)) {
            throw new Error("Device must follow Rechargeable contract");
        }
        
        if (device.getBatteryLevel() < 20) {
            console.log("Battery low, initiating recharge");
            device.recharge();
        } else {
            console.log(`Battery sufficient: ${device.getBatteryLevel()}%`);
        }
    }
}

class EducationManager {
    // Only depends on Cognitive contract
    static conductTraining(learner) {
        if (!followsContract(learner, CognitiveContract)) {
            throw new Error("Learner must follow Cognitive contract");
        }
        
        console.log("Starting cognitive training session...");
        learner.think();
        learner.learn();
    }
}

class SocialCoordinator {
    // Only depends on Social contract
    static organizeSocialEvent(...participants) {
        console.log("Organizing social interaction...");
        participants.forEach(participant => {
            if (followsContract(participant, SocialContract)) {
                participant.socialize();
                participant.communicate();
            } else {
                console.log("Warning: Participant doesn't follow Social contract");
            }
        });
    }
}

class BiologicalCareProvider {
    // Only depends on Biological contract
    static provideCare(organism) {
        if (!followsContract(organism, BiologicalContract)) {
            throw new Error("Organism must follow Biological contract");
        }
        
        console.log("Providing biological care...");
        organism.eat();
        organism.sleep();
        organism.breathe();
    }
}

/**
 * Demonstrates proper Interface Segregation in JavaScript
 */
function demonstrateISPCorrect() {
    console.log("=== Interface Segregation Principle - CORRECT (JavaScript) ===");
    console.log("");
    
    // Create different types of workers
    const alice = new Person("Alice");
    const factoryBot = new IndustrialRobot("FactoryBot-3000");
    const android = new HumanoidRobot("Android-Sara");
    
    console.log("1. Analyzing contract compliance:");
    console.log(`Alice follows contracts: ${getFollowedContracts(alice).join(', ')}`);
    console.log(`FactoryBot follows contracts: ${getFollowedContracts(factoryBot).join(', ')}`);
    console.log(`Android follows contracts: ${getFollowedContracts(android).join(', ')}`);
    
    console.log("");
    console.log("2. Work Management (only needs Workable contract):");
    WorkManager.assignTask(alice, "Design new product");
    WorkManager.assignTask(factoryBot, "Assembly line task");
    WorkManager.assignTask(android, "Customer service");
    
    console.log("");
    console.log("3. Managing workforce (polymorphic with Workable):");
    WorkManager.manageWorkforce(alice, factoryBot, android);
    
    console.log("");
    console.log("4. Maintenance (only for Mechanical devices):");
    // Note: Cannot pass alice here - she doesn't follow Mechanical contract!
    MaintenanceManager.performMaintenance(factoryBot);
    MaintenanceManager.performMaintenance(android);
    
    console.log("");
    console.log("5. Battery Management (only for Rechargeable devices):");
    // Note: Cannot pass alice here - she doesn't follow Rechargeable contract!
    MaintenanceManager.manageCharge(factoryBot);
    MaintenanceManager.manageCharge(android);
    
    console.log("");
    console.log("6. Cognitive Training (for thinking entities):");
    EducationManager.conductTraining(alice);
    EducationManager.conductTraining(android);  // Humanoid can think too!
    // Note: Cannot pass factoryBot - it doesn't follow Cognitive contract!
    
    console.log("");
    console.log("7. Social Coordination (for social entities):");
    SocialCoordinator.organizeSocialEvent(alice, android);
    // Note: Cannot pass factoryBot - it doesn't follow Social contract!
    
    console.log("");
    console.log("8. Biological Care (only for living organisms):");
    BiologicalCareProvider.provideCare(alice);
    // Note: Cannot pass robots - they don't follow Biological contract!
    
    console.log("");
    console.log("9. Network Operations (only for NetworkEnabled):");
    if (followsContract(android, NetworkEnabledContract)) {
        android.connectToNetwork();
        android.sendData("Status update");
        android.receiveData();
    }
    
    console.log("");
    console.log("=== Benefits of Interface Segregation ===");
    console.log("✓ Clients only depend on methods they actually use");
    console.log("✓ No forced implementation of irrelevant methods");
    console.log("✓ High cohesion within each contract");
    console.log("✓ Easy to extend with new capabilities");
    console.log("✓ Better testability and maintainability");
    console.log("✓ Duck typing works naturally with segregated contracts");
    
    console.log("");
    console.log("=== Contract Composition Examples ===");
    console.log("Person implements: Workable + Biological + Cognitive + Social + Physical");
    console.log("IndustrialRobot implements: Workable + Mechanical + Rechargeable + Programmable");
    console.log("HumanoidRobot implements: Workable + Cognitive + Social + Physical + Mechanical + Rechargeable + Programmable + NetworkEnabled");
    
    console.log("");
    console.log("10. Testing contract violations:");
    try {
        // This should fail - factoryBot doesn't follow Cognitive contract
        EducationManager.conductTraining(factoryBot);
    } catch (error) {
        console.log(`Expected error: ${error.message}`);
    }
    
    try {
        // This should fail - android doesn't follow Biological contract
        BiologicalCareProvider.provideCare(android);
    } catch (error) {
        console.log(`Expected error: ${error.message}`);
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        Person,
        IndustrialRobot,
        HumanoidRobot,
        WorkManager,
        MaintenanceManager,
        EducationManager,
        SocialCoordinator,
        BiologicalCareProvider,
        demonstrateISPCorrect,
        followsContract,
        getFollowedContracts,
        // Contract definitions
        WorkableContract,
        BiologicalContract,
        CognitiveContract,
        SocialContract,
        PhysicalContract,
        MechanicalContract,
        RechargeableContract,
        ProgrammableContract,
        NetworkEnabledContract
    };
}

// Run demonstration if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    demonstrateISPCorrect();
}